### PR TITLE
updated billing for base and gnosis!

### PR DESCRIPTION
**Pull request by memosys**
- updated watcher for base, gnosis
- can't add gnosis to swap as its not supported by 0x

### CODE REVIEW

File: web-portal/backend/src/utils/utils.service.ts [hunk 1]
```diff
@@ -168,39 +169,64 @@ export class UtilsService {
...
+      console.log('Applied New logs');
+    }
+  }
 
-    console.log('Applied New logs');
+  // Helper function to parse logs
+  parseLogs(logs: any[], network: string): IParsedLog[] {
+    return logs.map((log: any) => ({
+      tenantId: fromHex(log?.args?._identifier, 'string').replaceAll(`\x00`, ''),
+      amount: Number(log?.args?._amount),
```
Comment (plor) [inline web-portal/backend/src/utils/utils.service.ts:225]: This needs to be sized correctly. It is currently 18 decimals, will give the user too many credits (currently 1 trillion times too many).

File: web-portal/backend/src/utils/utils.service.ts [hunk 2]
```diff
@@ -168,39 +169,64 @@ export class UtilsService {
...
+    ];
+
+    // Fetch and parse logs for all clients
+    const allParsedLogs = await Promise.all(
+      clients.map(async ({ client, name }) => {
+        console.log(`Getting Event Logs from ${name}`);
+        const blockNumber = await client.getBlockNumber();
+        const logs = await client.getLogs({
+          event,
+          address: portrAddress,
```
Comment (plor) [inline web-portal/backend/src/utils/utils.service.ts:188]: This isn't guaranteed to be the same on every network.

File: web-portal/backend/src/utils/utils.service.ts [hunk 3]
```diff
@@ -168,39 +169,64 @@ export class UtilsService {
...
-      fromBlock: blockNumber - BigInt(1000),
-      toBlock: blockNumber,
-    });
+    console.log('Getting Event Logs from all clients');
+
+    // Define clients and their respective names
+    const clients = [
+      { client: opClient, name: 'optimism' },
+      { client: gnosisClient, name: 'gnosis' },
+      { client: baseClient, name: 'base' }
```
Comment (plor) [inline web-portal/backend/src/utils/utils.service.ts:178]: Is this manageable going forward. At what point do we have too many chains to support with the code this way?

**Review by plor — APPROVED** [review]
I don't like all of this stuff hardcoded this way. It feels like the chain data should be deployment specific. For example we don't have a way to support testnet tokens at this point.

File: services/postgres/schema.prisma [hunk 4]
```diff
@@ -100,6 +100,7 @@ model PaymentLedger {
   tenant          Tenant          @relation(fields: [tenantId], references: [id])
   tenantId        String
   referenceId     String          @unique
+  network         String?
```
Comment (plor) [inline services/postgres/schema.prisma:103]: Why?

File: web-portal/backend/src/utils/utils.service.ts [hunk 5]
```diff
@@ -168,39 +168,63 @@ export class UtilsService {
...
+      console.log('Applied New logs');
+    }
+  }
 
-    console.log('Applied New logs');
+  // Helper function to parse logs
+  parseLogs(logs: any[], network: string): IParsedLog[] {
+    return logs.map((log: any) => ({
+      tenantId: fromHex(log?.args?._identifier, 'string').replaceAll(`\x00`, ''),
+      amount: Number(log?.args?._amount * 10 ** -12),
```
Comment (plor) [inline web-portal/backend/src/utils/utils.service.ts:224]: A comment on this math might be useful. It looks like magic here.

**Review by plor — APPROVED** [review]
Looks good.

File: web-portal/backend/src/utils/utils.service.ts (hunk 5 above)
Comment (memosys) [reply web-portal/backend/src/utils/utils.service.ts:224]: added

**Pull request merged**